src/Superpower/Parse.cs

```diff
@@ -41,22 +41,29 @@ public static class Parse
             if (@operator == null) throw new ArgumentNullException(nameof(@operator));
             if (operand == null) throw new ArgumentNullException(nameof(operand));
             if (apply == null) throw new ArgumentNullException(nameof(apply));
-            return operand.Then(first => ChainOperatorRest(first, @operator, operand, apply));
-        }
 
-        static TextParser<T> ChainOperatorRest<T, TOperator>(
-            T firstOperand,
-            TextParser<TOperator> @operator,
-            TextParser<T> operand,
-            Func<TOperator, T, T, T> apply)
-        {
-            if (@operator == null) throw new ArgumentNullException(nameof(@operator));
-            if (operand == null) throw new ArgumentNullException(nameof(operand));
-            if (apply == null) throw new ArgumentNullException(nameof(apply));
-            return @operator.Then(opvalue =>
-                operand.Then(operandValue =>
-                    ChainOperatorRest(apply(opvalue, firstOperand, operandValue), @operator, operand, apply)))
-                    .Or(Return(firstOperand));
+            return input => {
+                var parseResult = operand(input);
+                if (!parseResult.HasValue) {
+                    return parseResult;
+                }
+
+                var result = parseResult.Value;
+
+                var operatorResult = @operator(parseResult.Remainder);
+                while (operatorResult.HasValue) {
+                    parseResult = operand(operatorResult.Remainder);
+
+                    if (!parseResult.HasValue) {
+                        return Result.CastEmpty<T, T>(parseResult);
```

Comment:
This is effectively a cast from `T` to `T` - I think we can just return `parseResult` here (and in the corresponding `TokenListParser` below).